File: 135-candy/135-candy.py
from heapq import *
class Solution(object):
    def candy(self, ratings):
        """
        :type ratings: List[int]
        :rtype: int
        """
        num = len(ratings)
        minimum = num
        count = [0] * num
        heap = []
        heapify(heap)
        for i, rating in enumerate(ratings):
            heappush(heap, [rating, i])
        seen = {}
        layer = 0
        prev = None
        while(heap):
            popped, index = heappop(heap)
            if index in seen:
                continue
            else:
                seen[index] = 1
                left, right = 0, 0
                if index + 1 in seen and ratings[index + 1] < popped:
                    right = count[index + 1] + 1
                if index - 1 in seen and ratings[index - 1] < popped:
                    left = count[index - 1] + 1
                count[index] = max([left, right, layer])
            if prev and prev < popped:
                layer += 1
        return sum(count) + minimum
            
            
        
            
        

File: 135-candy/135-candy.java
class Solution {
    public int candy(int[] ratings) {
        int n = ratings.length;
        int[] nums = new int[n];
        
        //left
        nums[0] = 1;
        for(int i = 1; i < n; i++){
            if(ratings[i] > ratings[i-1]){
                nums[i] = nums[i-1] + 1;
            }else if(ratings[i] <= ratings[i-1]){
                nums[i] = 1;
            }
        }
        //right
        for(int i = n - 2; i >= 0; i--){
            if(ratings[i] > ratings[i+1]){
                nums[i] = Math.max(nums[i],nums[i+1] + 1);
            }
        }
        int ans = 0;
        for(int num : nums){
            ans += num;
        }
        return ans;
        
        
    }

    
    
}